src/components/TreesMap/3d/trees-3d-layer.ts

```diff
@@ -0,0 +1,185 @@
+  return {
+    type: 'geojson',
+    data: {
+      features: [
+        {
+          type: 'Feature',
+          properties: { id: featureId },
+          geometry: {
+            type: 'Point',
+            coordinates: [lng, lat],
+          },
+          id: featureId,
+        },
+      ],
+      type: 'FeatureCollection',
+    },
+    promoteId: 'id',
+  };
+};
+
+export const trees3DHighlightLayer = {
+  id: trees3DHighlightLayerId,
+  //@ts-ignore
+  type: 'model',
+  source: trees3DHighlightSourceId,
+  layout: {
+    //@ts-ignore
+    'model-id': treeModelId,
+  },
+  paint: {
+    //@ts-ignore
+    'model-scale': [0.01, 0.01, 0.008],
+    'model-translation': [0, 0, 0],
+    'model-opacity': 1,
+  },
+};
+
+export const add3dHighlightLayer = function (
+  map: mapboxgl.Map,
+  feature: mapboxgl.MapboxGeoJSONFeature
+) {
+  const treeId = feature.id as string;
+  var featureTree = map.queryRenderedFeatures(undefined, {
+    layers: ['trees'],
+    filter: ['==', 'id', treeId],
+  })[0];
+
+  if (!featureTree) return;
+
+  //@ts-ignore
+  const lat = featureTree.geometry.coordinates[1];
+  //@ts-ignore
+  const lng = featureTree.geometry.coordinates[0];
+
+  if (!map.getSource(trees3DHighlightSourceId)) {
+    map.addSource(
+      trees3DHighlightSourceId,
+      //@ts-ignore
+      trees3DHighlightSource(lat, lng, treeId)
+    );
+    //@ts-ignore
+    map.addLayer(trees3DHighlightLayer);
+    map.moveLayer(trees3DHighlightLayerId, trees3DLayerId);
+  }
+
+  map.setFeatureState(
+    {
+      source: trees3DCylinderSourceId,
+      sourceLayer: 'tree_cylinders',
+      id: treeId,
+    },
+    { hovered: true }
+  );
+};
+
+export const remove3dHighlightLayer = function (
+  map: mapboxgl.Map,
+  treeId: string
+) {
+  map.setFeatureState(
+    {
+      source: trees3DCylinderSourceId,
+      sourceLayer: 'tree_cylinders',
+      id: treeId,
+    },
+    { hovered: false }
+  );
+  if (map.getSource(trees3DHighlightSourceId)) {
+    map.removeLayer(trees3DHighlightLayerId);
+    map.removeSource(trees3DHighlightSourceId);
+  }
+};
+
+export const add3dTreesCylinderMouseMoveListener = function (
+  map: mapboxgl.Map,
+  lastHoveredTreeIdRef: React.MutableRefObject<string | undefined>,
+  hoverCallback: (treeId: string) => void
+) {
+  map.on('mousemove', trees3DCylinderLayerId, function (e) {
+    const firstFeature = e.features ? e.features[0] : null;
+    if (firstFeature) {
+      const treeId = firstFeature.id as string;
+      if (
+        lastHoveredTreeIdRef.current &&
+        lastHoveredTreeIdRef.current !== treeId
+      ) {
+        remove3dHighlightLayer(map, lastHoveredTreeIdRef.current);
+      }
+      add3dHighlightLayer(map, firstFeature);
+      hoverCallback(treeId);
+    }
+  });
+};
+
+export const add3dTreesCylinderMouseLeaveListener = function (
+  map: mapboxgl.Map,
+  lastHoveredTreeIdRef: React.MutableRefObject<string | undefined>,
+  selectedTreeCylinderIdRef: React.MutableRefObject<string | undefined>,
```

Comment:
Aren't ref objects null when not there?

Reply:
Not sure, it was already used inconsistently in the codebase, e.g.

`useRef<string | undefined>(selectedTreeId);` and `const [hoveredTreeId, setHoveredTreeId] = useState<string | null>(null);`

I don't think it makes a difference, but if you know what is correct, I can change it.